enterprise/tailnet/coordinator_test.go

```diff
@@ -0,0 +1,260 @@
+package tailnet_test
+
+import (
+	"net"
+	"testing"
+
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"cdr.dev/slog/sloggers/slogtest"
+
+	"github.com/coder/coder/coderd/database"
+	"github.com/coder/coder/enterprise/tailnet"
+	agpl "github.com/coder/coder/tailnet"
+	"github.com/coder/coder/testutil"
+)
+
+func TestCoordinatorSingle(t *testing.T) {
+	t.Parallel()
+	t.Run("ClientWithoutAgent", func(t *testing.T) {
+		t.Parallel()
+		coordinator, err := tailnet.NewHACoordinator(slogtest.Make(t, nil), database.NewPubsubInMemory())
+		require.NoError(t, err)
+		defer coordinator.Close()
+
+		client, server := net.Pipe()
+		sendNode, errChan := agpl.ServeCoordinator(client, func(node []*agpl.Node) error {
+			return nil
+		})
+		id := uuid.New()
+		closeChan := make(chan struct{})
+		go func() {
+			err := coordinator.ServeClient(server, id, uuid.New())
+			assert.NoError(t, err)
+			close(closeChan)
+		}()
+		sendNode(&agpl.Node{})
+		require.Eventually(t, func() bool {
+			return coordinator.Node(id) != nil
+		}, testutil.WaitShort, testutil.IntervalFast)
+
+		err = client.Close()
+		require.NoError(t, err)
+		<-errChan
+		<-closeChan
+	})
+
+	t.Run("AgentWithoutClients", func(t *testing.T) {
+		t.Parallel()
+		coordinator, err := tailnet.NewHACoordinator(slogtest.Make(t, nil), database.NewPubsubInMemory())
```

Comment:
We should probably test this against multiple coordinators working together, otherwise it's not really replicating a deployment scenario.

Reply:
I have the HA tests below